rir/src/compiler/opt/gvn.cpp

```diff
@@ -0,0 +1,204 @@
+                return nextNumber;
+            };
+
+            auto i = Instruction::Cast(v);
+            if (!i) {
+                assignNumber(v);
+                return;
+            }
+
+            if (!i->producesRirResult())
+                return;
+
+            auto computeNumber = [&](Instruction* i) {
+                size_t klassNumber = i->gvnBase();
+
+                std::vector<size_t> args;
+                if (auto phi = Phi::Cast(i)) {
+                    bool success = false;
+                    auto res = computePhiNr(phi, success);
+                    if (success) {
+                        changed = true;
+                        number[phi] = res;
+                        if (!firstValue.count(res))
+                            firstValue[res] = phi;
+                    }
+                    return;
+                }
+
+                bool success = true;
+                i->eachArg([&](Value* a) {
+                    if (success && number.count(a))
+                        klassNumber = hash_combine(klassNumber, number.at(a));
+                    else
+                        success = false;
+                });
+                if (success) {
+                    i->eachArg([&](Value* a) { args.push_back(number.at(a)); });
+
+                    if (!classes.count(klassNumber) ||
+                        classes.at(klassNumber) != args) {
+                        while (classes.count(klassNumber))
+                            klassNumber++;
+                        storeNumber(i, klassNumber, args);
+                    } else {
+                        number[i] = klassNumber;
+                    }
+                    changed = true;
+                }
+            };
+
+            if (auto ld = LdConst::Cast(i)) {
+                SEXP constant = ld->c();
+                for (auto& c : constants) {
+                    if (R_compute_identical(c.first, constant, 0)) {
+                        number[i] = c.second;
+                        changed = true;
+                        return;
+                    }
+                }
+                while (classes.count(nextNumber))
+                    nextNumber++;
+                constants.emplace(constant, nextNumber);
+                storeNumber(i, nextNumber, {nextNumber});
+            } else {
+                // We allow instructions with those effects to be deduplicated.
+                auto effects =
+                    i->effects & (~(Effects(Effect::Error) | Effect::Warn |
+                                    Effect::Visibility | Effect::Force));
+                if (effects.empty())
+                    computeNumber(i);
+                else
+                    assignNumber(i);
+            }
+        };
+
+        while (changed) {
+            changed = false;
+            DominatorTreeVisitor<>(dom).run(cls->entry, [&](BB* bb) {
+                for (auto i : *bb)
+                    computeGN(i);
```

Comment:
If it is a dominator tree, why would be need to iterate?

Reply:
as you can see on a simple if-then-else example, dominator tree order does not guarantee that all inputs are already assigned a number. tbh, I am not even sure if it is faster than depth first.